### ISSUE TITLE
Use of PACKER_KEY_INTERVAL is inconsistent between builders

### ISSUE THREAD

**Issue by andrewsav-datacom** [hashicorp/packer#6496]
- Packer version from `packer version`:  1.2.4
- Host platform: Windows Server 2016
- Debug log output: at this point irrelevant, can be provided upon request
- The _simplest example template and scripts_ needed to reproduce the bug : at this point irrelevant, can be provided upon request

[Virtualbox builder documentation](https://www.packer.io/docs/builders/virtualbox-iso.html#boot-command) says:

>The boot command is sent to the VM through the VBoxManage utility in as few invocations as possible. We send each character in groups of 25, with a default delay of 100ms between groups. The delay alleviates issues with latency and CPU contention. If you notice missing keys, you can tune this delay by specifying e.g. PACKER_KEY_INTERVAL=500ms to wait longer between each group of characters.

[VmWare builder documentation](https://www.packer.io/docs/builders/vmware-iso.html#boot_command) says:

>Keystrokes are typed as separate key up/down events over VNC with a default 100ms delay. The delay alleviates issues with latency and CPU contention. For local builds you can tune this delay by specifying e.g. PACKER_KEY_INTERVAL=10ms to speed through the boot command.

The key difference here is that PACKER_KEY_INTERVAL is inserted after every 25 characters on Virtualbox and after every single character on VmWare.

I have a packer template that builds both VmWare and Virtualbox images in parallel. If I set PACKER_KEY_INTERVAL to 100s, keystrokes get lost on Virtualbox. If I set it to 500s it is so painfully slow on VmWare.

It does not appear that I can configure those intervals separately for the builders. It also appears that the two builders uses the same parameter very differently.

Can that be fixed please? Preferably a setting per builder, maybe in the template file itself but environment would do too. If not then at least make them consistent so they both work after a single character not after 25.

Thank you in advance!

**Comment by SwampDragons** [comment]
Thanks for opening this issue. This is definitely a bad user experience that we should fix.

**Comment by rickard-von-essen** [comment]
There is a technical difference in how keystrokes are passed. Hence the long delay between _groups_ for VirtualBox and short delays between each key in VMware.

I think we should just postfix the environment variable with the builder type, e.g. `PACKER_KEY_INTERVAL_VIRTUALBOX`.

**Comment by AndrewSav** [comment]
That would work for me, thanks!

**Comment by mwhooker** [comment]
echoing everyone else, this is a good point. thanks for opening the ticket! I'm thinking it might be best to add a config option to each builder, like `boot_command_interval` which would allow you to fine tune it between builders.

I agree it's a bit weird that we use the same construct to set delays between keystroke in vnc and groups in pcxt, but I think we can get past it.

I want to trend us away from configuring packer through the environment. Where we do it we're mostly trying to hide details from the user. I think we should keep `PACKER_KEY_INTERVAL` but override it if it's set in the builder.

What do people think about that?

**Comment by AndrewSav** [comment]
I'd rather have it in the config, then in environment variable although both works. It's better in config because this way you keep all configuration information in the same place and does not spread it between two location.

As far as overriding goes I do not have problems with this either as long as it's the config that overrides the environment variable, and not the other way around.

>I agree it's a bit weird that we use the same construct to set delays between keystroke in vnc and groups in pcxt, but I think we can get past it.

Finally, may be if they called differently to better reflect what they do all weirdness could be eliminated. Something like `boot_command__keystroke_interval` and `boot_command_group_interval`. Of course only one of the two would be valid for a particular builder.

**Comment by mwhooker** [comment]
Edited my comment above, but I'll repeat here. Another reason we like to use env vars is if we're exposing environment dependent config, which usually has to do with timing. This definitely falls into that category, and so I don't think adding to the json config is a good idea.

I think splitting these two delays into two env vars is the best solution

**Comment by AndrewSav** [comment]
Sorry, what is config option then? Command line?

**Comment by mwhooker** [comment]
I'm not sure I follow. I originally proposed adding it to the JSON, but changed my mind and now think they work best as env vars.

**Comment by AndrewSav** [comment]
Oh I see, it was not clear for me that you changed your mind. Now I understand.

You asked for opinion, got one that aligns with one you expressed and immediately changed yours to opposite. This is very frustrating, makes one wonder if their opinion that was explicitly asked for even matters.

To stay on a positive note - either will work for me json or environment. Thanks.

**Comment by mwhooker** [comment]
Sorry you were frustrated. Please know that the proposal is not personal, and is the result of careful consideration that included your input

**Comment by SwampDragons** [comment]
I'm torn on this one. My gut instinct is that `PACKER_KEY_INTERVAL` and `PACKER_GROUP_INTERVAL` are not obvious enough. The goal here is to abstract away things users shouldn't have to know, and in an ideal situation users wouldn't have to care whether the pause is between every key or a group of keys, or what the underlying mechanism for forwarding keypresses is. In that situation, I think @rickard-von-essen's suggestion feels like a better user experience... but it leads to an explosion of env vars and requires that users set a bunch of custom vars if they have several different builders. So maybe the `PACKER_KEY_INTERVAL` and `PACKER_GROUP_INTERVAL` are okay if we document it really well and make sure the defaults are good.

But I also think I like the idea of setting it in a config, even though we have a history of setting it as an environment variable; if a user has a need to run a config on multiple environments that benefit from different intervals, they can always set a config value using an environment variable because we have a templating engine for that. And I am all for making things explicit and obvious, which having a config option clearly is.

**Comment by SwampDragons** [comment]
I've taken a stab at this in PR #6616; I've attached a linux binary to that PR if you want to give it a whirl. 

tl;dr: I've created two new packer config options, `boot_key_interval`, and `boot_keygroup_interval` depending on whether the builder's bootcommand driver groups the keys or pauses between each.  Check out the updated docs on the PR for more details and a usage example for each builder.

**Comment by AndrewSav** [comment]
@SwampDragons I'm running most of my builds on windows. It would be possible for me to test the linux binary but that will take a bit more time. As it stands right now, linux or windows I probably won't be able to get to it this week. Thank you for the PR!

**Comment by SwampDragons** [comment]
no problem. I'll add a windows binary to the PR, and you can test it when you can.
